### PR TITLE
[dbnode] Use go context to cancel index query workers after timeout

### PR DESCRIPTION
**Pull request by ryanhall07**
This prevents a race in the previous code that only checked the timeout
when scheduling a new worker. Once a worker was dispatched it could in
theory run forever since it would not recheck the timeout.

**What this PR does / why we need it**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/dbnode/storage/index_query_concurrent_test.go [hunk 1]
```diff
@@ -254,7 +253,14 @@ func testNamespaceIndexHighConcurrentQueries(
 						r index.QueryResults,
 						logFields []opentracinglog.Field,
 					) (bool, error) {
-						timeoutWg.Wait()
+						// block until the query will be canceled due to timeout.
```
Comment (ryanhall07) [inline src/dbnode/storage/index_query_concurrent_test.go:256]: @robskillington this test broke because it was relying on the original flow control, that has now changed. now there is a no longer a goroutine waiting for the timeout and returning. instead there is a goroutine watching for the timeout and interrupting the workers (which ultimately results in a return). however, that doesn't work with the previous flow since `block.Query` never runs to get interrupted.

this patch makes the test go green, but unclear if it has the same coverage as before. I'm not sure what the previous test was actually trying to accomplish. It appears it was trying to stress test racing to close the ctx. However, not even sure if that's possible anymore since we are properly interrupting the workers _before_ returning from the rpc call.

File: src/dbnode/storage/index.go [hunk 2]
```diff
@@ -1560,17 +1561,47 @@ func (i *nsIndex) queryWithSpan(
...
+	case <-stdCtx.Done():
+		return false, fmt.Errorf("caller canceled the query before it started")
+	default:
+	}
+
+	if timeout > 0 {
+		// launch a goroutine to interrupt the query workers after the timeout.
+		fnDoneCh := make(chan struct{}, 1) // don't block the defer in the case of a timeout.
+		defer func() {
+			fnDoneCh <- struct{}{}
```
Comment (ryanhall07) [inline src/dbnode/storage/index.go:1590]: this is why we really need to start embracing the std go ctx instead. ideally this would be

```
ctx, cancel := WithCancel(ctx)
defer cancel()

select {
   <-ctx.Done()
}
```

however, the m3 `Context` makes that difficult to achieve, since it's doesn't have the same immutability pattern. If you do `Context.SetGoContext(ctx)` you'll impact the ctx of the caller with the new cancel, which is not good.

File: src/dbnode/storage/index.go [hunk 3]
```diff
@@ -1594,89 +1625,62 @@ func (i *nsIndex) queryWithSpan(
 			break
 		}
 
-		if applyTimeout := timeout > 0; !applyTimeout {
+		wg.Add(1)
```
Comment (ryanhall07) [inline src/dbnode/storage/index.go:1628]: this could use some close eyes. it should behave exactly as before. instead of launching a goroutine to watch the timeout and return, we are using a goroutine to watch the timeout and interrupt. when the workers are interrupted, it will ultimately return preserving the existing behavior and properly cleaning up.

File: src/dbnode/storage/index_test.go [hunk 4]
```diff
@@ -378,6 +381,65 @@ func TestNamespaceIndexQueryNoMatchingBlocks(t *testing.T) {
...
+			q index.Query,
+			opts index.QueryOptions,
+			r index.QueryResults,
+			logFields []opentracinglog.Field,
+		) (bool, error) {
+			for {
+				if !c.TryCheckout() {
+					return false, index.ErrCancelledQuery
+				}
+				time.Sleep(time.Millisecond)
```
Comment (wesleyk) [inline src/dbnode/storage/index_test.go:418]: why is this needed?

**Review by wesleyk — APPROVED** [review]
LGTM

File: src/dbnode/storage/index_test.go (hunk 4 above)
Comment (ryanhall07) [reply src/dbnode/storage/index_test.go:418]: figured it was better than tight looping in a test.

Comment (ryanhall07) [reply src/dbnode/storage/index_test.go:418]: this runs in a separate goroutine (from the worker pool), so probably best to not crush the cpu.

File: src/dbnode/storage/index.go [hunk 5]
```diff
@@ -1560,17 +1561,47 @@ func (i *nsIndex) queryWithSpan(
...
+	// always cancel to cleanup. this might be a no-op if it was already canceled by the background task to interrupt
+	// the workers.
 	defer cancellable.Cancel()
 
-	indexMatchingStartTime := time.Now()
-	var totalWaitTime time.Duration
+	stdCtx, ctxFound := ctx.GoContext()
+	if !ctxFound {
+		stdCtx = stdctx.Background()
+	}
```
Comment (robskillington) [inline src/dbnode/storage/index.go:1579]: Should we add a convenience method to the M3 context such as:
```go
goCtx := ctx.GoContextOrBackground()
```

File: src/dbnode/storage/index.go [hunk 6]
```diff
@@ -1560,17 +1561,47 @@ func (i *nsIndex) queryWithSpan(
...
+	select {
+	case <-stdCtx.Done():
+		return false, fmt.Errorf("caller canceled the query before it started")
```
Comment (robskillington) [inline src/dbnode/storage/index.go:1582]: nit: Use a `var errQueryCancelledBeforeStarted = errors.New("....")` to the top of this file if this is a constant?

File: src/dbnode/storage/index.go (hunk 2 above)
Comment (robskillington) [reply src/dbnode/storage/index.go:1590]: Yeah maybe we should honestly just remove the whole concept of "not having a timeout" here and rely on the context having an existing one?

As @nbroyles discovered it's basically guaranteed you'll have a serverside Go context with a timeout specified already with TChannel Thrift.

File: src/dbnode/storage/index.go (hunk 3 above)
Comment (robskillington) [reply src/dbnode/storage/index.go:1584]: Yeah I think this looks good.

File: src/dbnode/storage/index.go [hunk 7]
```diff
@@ -1688,12 +1628,12 @@ func (i *nsIndex) queryWithSpan(
...
 	i.metrics.querySearchTime.ByDocs.Record(results.TotalDocsCount(), results.TotalDuration().Search)
 
-	return exhaustive, nil
+	return exhaustive, err
 }
 
 func (i *nsIndex) execBlockQueryFn(
 	ctx context.Context,
-	cancellable *xresource.CancellableLifetime,
+	_ *xresource.CancellableLifetime,
```
Comment (robskillington) [inline src/dbnode/storage/index.go:1636]: Should we just remove these from the exec block query fn's since they are unused now? Or is this just here until aggregate does not use cancellable anymore?

File: src/dbnode/storage/index.go [hunk 8]
```diff
@@ -1740,7 +1680,7 @@ func (i *nsIndex) execBlockQueryFn(
 
 func (i *nsIndex) execBlockWideQueryFn(
 	ctx context.Context,
-	cancellable *xresource.CancellableLifetime,
+	_ *xresource.CancellableLifetime,
```
Comment (robskillington) [inline src/dbnode/storage/index.go:1683]: Should we just remove these from the exec block query fn's since they are unused now? Or is this just here until aggregate does not use cancellable anymore?

File: src/dbnode/storage/index/block.go [hunk 9]
```diff
@@ -429,13 +429,9 @@ func (b *block) Query(
...
 func (b *block) queryWithSpan(
 	ctx context.Context,
-	cancellable *xresource.CancellableLifetime,
 	query Query,
 	opts QueryOptions,
-	results DocumentResults,
-	sp opentracing.Span,
-	logFields []opentracinglog.Field,
-) (bool, error) {
+	results DocumentResults) (bool, error) {
```
Comment (robskillington) [inline src/dbnode/storage/index/block.go:434]: nit: For consistency the majority of our indentation for function definitions usually has each argument with a trailing comma like this was before:
e.g.
```go
func (b *block) queryWithSpan(
  ctx context.Context,
  ...
  results DocumentResults,
) (bool, error) {
  // ...
}
```

File: src/dbnode/storage/index/block.go [hunk 10]
```diff
@@ -498,6 +486,13 @@ func (b *block) queryWithSpan(
 		docsPool.Put(batch)
 	}()
 
+	select {
+	case <-ctx.MustGoContext().Done():
+		// indexNs will log something useful.
+		return false, ErrCancelledQuery
+	default:
+	}
```
Comment (robskillington) [inline src/dbnode/storage/index/block.go:494]: If we already check for cancellation at the query worker pool level before the start of this query starts executing, do we need one as well before the batches do you think?

Probably fine to be here, but also probably would be fine to omit this too since we check it before we get the token and so we probably momentarily check this prior to now (haven't started searches yet I believe?).

File: src/x/sync/types.go [hunk 11]
```diff
@@ -88,6 +89,9 @@ type WorkerPool interface {
 
 	// GoWithTimeoutInstrument instruments GoWithTimeout with timing information.
 	GoWithTimeoutInstrument(work Work, timeout time.Duration) ScheduleResult
+
+	// GoWithCtx waits until a worker is available or the provided ctx is canceled.
+	GoWithCtx(ctx context.Context, work Work) ScheduleResult
```
Comment (robskillington) [inline src/x/sync/types.go:94]: nit: Other naming usually only uses abbreviations in local vars or struct fields and the type itself is fully spelled out. Guess you could call it Java-iesque naming... but anyhow would be slightly more consistent to call this `GoWithContext(...)`.

File: src/x/sync/worker_pool.go [hunk 12]
```diff
@@ -104,3 +105,25 @@ func (p *workerPool) GoWithTimeoutInstrument(work Work, timeout time.Duration) S
...
 	}
 }
+
+func (p *workerPool) GoWithCtx(ctx context.Context, work Work) ScheduleResult {
+	// Don't give out a token if the ctx has already been canceled.
+	select {
+	case <-ctx.Done():
+		return ScheduleResult{Available: false, WaitTime: 0}
+	default:
+	}
```
Comment (robskillington) [inline src/x/sync/worker_pool.go:115]: 👍

**Review by robskillington — COMMENTED** [review]
Looks good to me as is

File: src/dbnode/storage/index.go (hunk 5 above)
Comment (ryanhall07) [reply src/dbnode/storage/index.go]: I went with `ctx.MustGoContext() stdctx.Context` since it is guaranteed to be set for a thrift rpc handler. Also added a new helper constructors for tests `NewBackground()` and `NewWithGoContext()`

File: src/dbnode/storage/index.go (hunk 6 above)
Comment (ryanhall07) [reply src/dbnode/storage/index.go]: just used the standard `index.ErrQueryCanceled`

File: src/dbnode/storage/index.go (hunk 7 above)
Comment (ryanhall07) [reply src/dbnode/storage/index.go:1636]: removed since I updated the aggregate path as well

File: src/dbnode/storage/index/block.go (hunk 10 above)
Comment (ryanhall07) [reply src/dbnode/storage/index/block.go:494]: yea I added this to get a concurrency test to actually return a timeout, since the test had less blocks than the batch :/  not a good reason. I'll remove and get the test working another way. 

it might not be a bad idea to check inside the iterator when the len(batch) == 0, since that would mean search has run

File: src/dbnode/client/fetch_tagged_results_accumulator_merge_test.go [hunk 13]
```diff
@@ -592,7 +592,7 @@ func (td testDatapoints) toRPCSegments(th testFetchTaggedHelper, start time.Time
 	for _, dp := range td {
 		require.NoError(th.t, enc.Encode(dp, testFetchTaggedTimeUnit, nil), fmt.Sprintf("%+v", dp))
 	}
-	reader, ok := enc.Stream(context.NewContext())
+	reader, ok := enc.Stream(context.NewBackground())
```
Comment (ryanhall07) [inline src/dbnode/client/fetch_tagged_results_accumulator_merge_test.go:595]: This is a mechanical refactoring that I scoped to test files only

```
git grep --name-only "\.NewContext()"  | grep "_test.go" | xargs sed -i '' 's/\.NewContext()/\.NewBackground()/'
```

It was getting painful to update individual failing tests that were missing a go context when querying
